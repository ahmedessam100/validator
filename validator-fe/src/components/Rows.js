import React, { Component } from 'react';

class Rows extends Component {
    render() {
        return (
            <tr style={{textAlign: "center"}}>
                <td style={{textAlign: "center"}} id="name">{this.props.customerName}</td>
                <td style={{textAlign: "center"}} id="country">{this.props.country}</td>
                <td style={{textAlign: "center"}} id="code">{this.props.code}</td>
                <td style={{textAlign: "center"}} id="phone">{this.props.phone}</td>
                <td style={{textAlign: "center"}} id="state">{this.props.state}</td>
            </tr>
        );
    }
}

export default Rows;
