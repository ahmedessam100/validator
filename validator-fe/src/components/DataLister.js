import React, { Component } from 'react';
import Rows from './Rows';

class DataLister extends Component {

    constructor(props) {
        super(props);
        this.state = {
            dataReady: false,
            data: [],
            originalData: [],
            filterKey: '',
            selection: '',
            error: null
        };
    }

    fetchData = () => {
        fetch('http://localhost:8080/getData')
            .then((res) => res.json())
            .then((res) => this.setState({
                data: res.reduce((customers, currCustomer) => {

                    let customer = {
                        name: currCustomer.name,
                        phoneNumber: currCustomer.phoneNumber,
                        code: currCustomer.code,
                        country: currCustomer.country,
                        validation: currCustomer.state
                    };

                    customers.push(customer);

                    return customers;
                }, []),

                dataReady: true
            }))
            .then(() => {
                this.setState({ originalData: [...this.state.data] } )
            })
            .catch((e) => this.setState({ error: e, dataReady: false }));
    };

    fetchDataByCountry = () => {

        fetch('http://localhost:8080/getDataByCountry/' + this.state.filterKey)
            .then((res) => res.json())
            .then((res) => this.setState({
                data: res.reduce((customers, currCustomer) => {

                    let customer = {
                        name: currCustomer.name,
                        phoneNumber: currCustomer.phoneNumber,
                        code: currCustomer.code,
                        country: currCustomer.country,
                        validation: currCustomer.state
                    };

                    customers.push(customer);

                    return customers;
                }, []),

                dataReady: true
            }))
            .catch((e) => this.setState({ error: e, dataReady: false }));
    };

    fetchDataByState = () => {

        fetch('http://localhost:8080/getDataByState/' + this.state.filterKey)
            .then((res) => res.json())
            .then((res) => this.setState({
                data: res.reduce((customers, currCustomer) => {

                    let customer = {
                        name: currCustomer.name,
                        phoneNumber: currCustomer.phoneNumber,
                        code: currCustomer.code,
                        country: currCustomer.country,
                        validation: currCustomer.state
                    };

                    customers.push(customer);

                    return customers;
                }, []),

                dataReady: true
            }))
            .catch((e) => this.setState({ error: e, dataReady: false }));
    };


    filterData = (e) => {
        if(this.state.selection === 'country')
        {
            this.fetchDataByCountry();
        }
        else if(this.state.selection === 'validation')
        {
            this.fetchDataByState();
        }
    };

    handleChange = (e) => {

        if (e.target.id === 'filterSelection')
        {
            this.setState({ selection: e.target.value });
        }
        else if (e.target.id === 'searchKey')
        {
            if (e.target.value === '')
                this.reset();

            this.setState({ filterKey : e.target.value });
        }
    };

    reset = () => {
        /* reset the data */
        this.setState({
            data: [...this.state.originalData]
        })
    };

    componentDidMount() {
        this.fetchData();
    }

    render() {

        /* If the data is not loaded */
        if (!this.state.dataReady || this.state.error)
        {
            return this.state.error ? <p>{this.state.error.message}</p> : <div />
        }

        /* Filling the table with rows */
        const rows: JSX.Element[] = [];
        rows.push(<tr>
                        <th style={{textAlign: "center"}}>Name</th>
                        <th style={{textAlign: "center"}}>Country</th>
                        <th style={{textAlign: "center"}}>Code</th>
                        <th style={{textAlign: "center"}}>Phone</th>
                        <th style={{textAlign: "center"}}>State</th>
                  </tr>);

        let customer;

        for (let i = 0; i < this.state.data.length; i++)
        {
            customer = this.state.data[i];

            rows.push(<Rows customerName={customer.name} phone={customer.phoneNumber} country={customer.country} code={'+' + customer.code} state={customer.validation}/>)
        }

        return (
                <div style={{overflow:"auto"}} >
                    <label style={{paddingLeft: 900, fontSize: 30}}>Filter</label>

                    <select id={"filterSelection"} onChange={ this.handleChange }>
                        <option value="" selected disabled hidden>Select</option>
                        <option value={"country"}>Country</option>
                        <option value={"validation"}>State</option>
                    </select>

                    <input type="text" id="searchKey" name="key" onChange={ this.handleChange }/>
                    <input type={"submit"} id={"filter"} value={"Filter"} onClick={ this.filterData }/>
                    <input type={"submit"} id={"reset"}  value={"Reset"} onClick={ this.reset }/>

                    <table>
                        {rows}
                    </table>
                </div>
        );
    }
}

export default DataLister;
