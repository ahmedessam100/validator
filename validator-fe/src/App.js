import { BrowserRouter, Route } from 'react-router-dom';
import DataLister from './components/DataLister';

function App() {
  return (
    <BrowserRouter>
      <div className="App">
        <Route exact path='/' component={DataLister}/>
      </div>
    </BrowserRouter>
  );
}

export default App;
