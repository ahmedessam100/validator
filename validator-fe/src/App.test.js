import App from './App';
import React from "react";
import { shallow, mount } from "enzyme";
import DataLister from "./components/DataLister";
import Rows from "./components/Rows";
import {render, fireEvent, cleanup} from '@testing-library/react';
import toJson from "enzyme-to-json";

afterEach(cleanup);

const customer = {
  name: 'test',
  country: 'Morocco',
  code: '212',
  phone: '6007989253',
  phoneNumber: '(212) 6007989253',
  validation: 'not valid',
  regex: '\\(212\\)\\ ?[5-9]\\d{8}$'
};

describe('Testing the components to render without crashing', () => {
  it('renders without crashing', () => {
    shallow(<App/>);
  });

  it('renders data component without crashing', () => {
    shallow(<DataLister />);
  });

  it('renders rows component without crashing', () => {
    shallow(<Rows />);
  });
});

it("DataLister component renders correctly with no error message", () => {
  const wrapper = mount(<DataLister />);
  expect(wrapper.state("error")).toEqual(null);
});

describe('Testing data transfer between DataLister and Rows components', () => {

  it('accepts customer props', () => {
    const wrapper = mount(<Rows customerName={customer.name} phone={customer.phone} country={customer.country} code={customer.code} state={customer.state} />);
    expect(wrapper.props().customerName).toEqual(customer.name);
    expect(wrapper.props().phone).toEqual(customer.phone);
    expect(wrapper.props().country).toEqual(customer.country);
    expect(wrapper.props().code).toEqual(customer.code);
    expect(wrapper.props().state).toEqual(customer.state);
  });

  it('contains customer data', () => {
    const wrapper = mount(<Rows customerName={customer.name} phone={customer.phone} country={customer.country} code={customer.code} state={customer.state}/>);
    expect(wrapper.find({id: "name"}).getElement().props.children).toEqual(customer.name);
    expect(wrapper.find({id: "phone"}).getElement().props.children).toEqual(customer.phone);
    expect(wrapper.find({id: "country"}).getElement().props.children).toEqual(customer.country);
    expect(wrapper.find({id: "code"}).getElement().props.children).toEqual(customer.code);
    expect(wrapper.find({id: "state"}).getElement().props.children).toEqual(customer.state);
  });
});

it("renders correctly", () => {
  const tree = mount(<App />);
  expect(toJson(tree)).toMatchSnapshot();
});

it('phone number validation functioning correctly', () => {
  const wrapper = shallow(<DataLister />);
  let value = wrapper.instance().isValid(customer.phoneNumber, customer.code);
  expect(value).toEqual(customer.validation);
});
