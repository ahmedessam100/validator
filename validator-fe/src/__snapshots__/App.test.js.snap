// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<App>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <div
        className="App"
      >
        <Route
          component={[Function]}
          exact={true}
          path="/"
        >
          <DataLister
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
          >
            <div />
          </DataLister>
        </Route>
      </div>
    </Router>
  </BrowserRouter>
</App>
`;
